File: Java/Bsv.java
/* (C) Stefan John / Stenway / Stenway.com / 2023 */

import java.util.ArrayList;
import java.util.Objects;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CharsetEncoder;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;

class Bsv {
	static byte[] encodeBsv(String[][] jaggedArray) {
		ArrayList<byte[]> parts = new ArrayList<byte[]>();
		byte[] lineBreakByte = new byte[]{(byte)0xFF};
		byte[] valueSeparatorByte = new byte[]{(byte)0xFE};
		byte[] nullValueByte = new byte[]{(byte)0xFD};
		byte[] emptyStringByte = new byte[]{(byte)0xFC};
		CharsetEncoder encoder = StandardCharsets.UTF_8.newEncoder();
		boolean isFirstLine = true;
		int resultLength = 0;
		for (String[] line : jaggedArray) {
			if (isFirstLine == false) { parts.add(lineBreakByte); resultLength++; }
			isFirstLine = false;
			boolean isFirstValue = true;
			for (String value : line) {
				if (isFirstValue == false) { parts.add(valueSeparatorByte); resultLength++; }
				isFirstValue = false;
				if (value == null) { parts.add(nullValueByte); resultLength++; }
				else if (value.length() == 0) { parts.add(emptyStringByte); resultLength++; }
				else {
					byte[] valueBytes;
					try {
						ByteBuffer byteBuffer = encoder.encode(CharBuffer.wrap(value));
						valueBytes = new byte[byteBuffer.limit()];
						byteBuffer.get(valueBytes);
					} catch (Exception e) { throw new RuntimeException("Invalid string value", e); }
					parts.add(valueBytes); resultLength += valueBytes.length;
				}
			}
		}
		byte[] result = new byte[resultLength];		
		ByteBuffer resultBuffer = ByteBuffer.wrap(result);
		for (byte[] part : parts) { resultBuffer.put(part); }
		return result;
	}
	
	static String[][] decodeBsv(byte[] bytes) {
		CharsetDecoder decoder = StandardCharsets.UTF_8.newDecoder();
		decoder.onMalformedInput(CodingErrorAction.REPORT);
		decoder.onUnmappableCharacter(CodingErrorAction.REPORT);
		ByteBuffer byteBuffer = ByteBuffer.wrap(bytes);
		ArrayList<String[]> result = new ArrayList<String[]>();
		ArrayList<String> currentLine = new ArrayList<String>();
		int currentIndex = -1;
		for (;;) {
			int lastIndex = currentIndex;
			for (;;) {
				currentIndex++;
				if (currentIndex >= bytes.length) { currentIndex = -1; break; }
				if ((bytes[currentIndex] & 0xFF) >= 0xFE) { break; }
			}
			ByteBuffer valueBytes = ((ByteBuffer)(byteBuffer.duplicate().position(lastIndex+1).limit(currentIndex < 0 ? bytes.length : currentIndex))).slice();
			if (valueBytes.capacity() == 1 && valueBytes.get(0) == (byte)0xFD) { currentLine.add(null); }
			else if (valueBytes.capacity() == 1 && valueBytes.get(0) == (byte)0xFC) { currentLine.add(""); }
			else if (valueBytes.capacity() >= 1) {
				try { currentLine.add(decoder.decode(valueBytes).toString()); }
				catch (Exception e) { throw new RuntimeException("Invalid string value", e); }
			} else if ((((currentIndex >= 0 && bytes[currentIndex] == 0xFF) || (currentIndex < 0)) && ((lastIndex < 0) || (lastIndex >= 0 && bytes[lastIndex] == 0xFF))) == false) { throw new RuntimeException("Invalid BSV value byte sequence"); }
			if (currentIndex < 0) { break; }
			else if (currentIndex >= 0 && bytes[currentIndex] == (byte)0xFF) {
				result.add(currentLine.toArray(new String[0]));
				currentLine.clear();
			}
		}
		result.add(currentLine.toArray(new String[0]));
		return result.toArray(new String[0][]);
	}
	
	// ----------------------------------------------------------------------
	
	static void saveBsv(String[][] jaggedArray, String filePath) throws IOException {
		Objects.requireNonNull(filePath);
		Files.write(Paths.get(filePath), encodeBsv(jaggedArray));
	}
	
	static String[][] load(String filePath) throws IOException {
		Objects.requireNonNull(filePath);
		return decodeBsv(Files.readAllBytes(Paths.get(filePath)));
	}
	
	static void appendBsv(String[][] jaggedArray, String filePath) throws IOException {
		OutputStream stream;
		boolean existed;
		try {
			stream = Files.newOutputStream(Paths.get(filePath), StandardOpenOption.APPEND);
			existed = true;
		} catch (java.nio.file.NoSuchFileException exception) {
			stream = Files.newOutputStream(Paths.get(filePath), StandardOpenOption.CREATE_NEW);
			existed = false;
		}
		try {
			if (existed == true) { stream.write(0xFF); }
			stream.write(encodeBsv(jaggedArray));
		} finally { stream.close(); }
	}
	
	// ----------------------------------------------------------------------

	static String jaggedArrayToString(String[][] jaggedArray) {
		StringBuilder sb = new StringBuilder();
		sb.append("[");
		for (String[] line : jaggedArray) {
			sb.append("\n  [");
			boolean isFirst = true;
			for (String value : line) {
				if (!isFirst) { sb.append(", "); }
				isFirst = false;
				if (value == null) { sb.append("null"); }
				else { sb.append("\""+value+"\""); }
			}
			sb.append("]");
		}
		sb.append("\n]");
		return sb.toString();
	}
	
	public static void main(String[] args) {
		try {
			String[][] jaggedArray = {{"Hello", "\uD83C\uDF0E", null, ""}, {"Test\uD834\uDD1E"}};
			saveBsv(jaggedArray, "Test.bsv");
			String[][] loadedJaggedArray = load("Test.bsv");
			System.out.println(jaggedArrayToString(loadedJaggedArray));
			
			appendBsv(new String[][]{{"ABC"}}, "Append.bsv");
		} catch(Exception e) {
			System.out.println(e.toString());
		}
		
	}
}

File: Kotlin/bsv.kt
/* (C) Stefan John / Stenway / Stenway.com / 2023 */

import java.io.IOException
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.NoSuchFileException
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import java.util.*

fun encodeBsv(jaggedArray: Array<Array<String?>>): ByteArray {
	val parts = ArrayList<ByteArray>()
	val lineBreakByte = byteArrayOf(0xFF.toByte())
	val valueSeparatorByte = byteArrayOf(0xFE.toByte())
	val nullValueByte = byteArrayOf(0xFD.toByte())
	val emptyStringByte = byteArrayOf(0xFC.toByte())
	val encoder = StandardCharsets.UTF_8.newEncoder()
	var isFirstLine = true
	var resultLength = 0
	for (line in jaggedArray) {
		if (isFirstLine == false) {
			parts.add(lineBreakByte)
			resultLength++
		}
		isFirstLine = false
		var isFirstValue = true
		for (value in line) {
			if (isFirstValue == false) {
				parts.add(valueSeparatorByte)
				resultLength++
			}
			isFirstValue = false
			if (value == null) {
				parts.add(nullValueByte)
				resultLength++
			} else if (value.length == 0) {
				parts.add(emptyStringByte)
				resultLength++
			} else {
				var valueBytes: ByteArray
				try {
					val byteBuffer = encoder.encode(CharBuffer.wrap(value))
					valueBytes = ByteArray(byteBuffer.limit())
					byteBuffer[valueBytes]
				} catch (e: Exception) {
					throw RuntimeException("Invalid string value", e)
				}
				parts.add(valueBytes)
				resultLength += valueBytes.size
			}
		}
	}
	val result = ByteArray(resultLength)
	val resultBuffer = ByteBuffer.wrap(result)
	for (part in parts) {
		resultBuffer.put(part)
	}
	return result
}

fun decodeBsv(bytes: ByteArray): Array<Array<String?>> {
	val decoder = StandardCharsets.UTF_8.newDecoder()
	decoder.onMalformedInput(CodingErrorAction.REPORT)
	decoder.onUnmappableCharacter(CodingErrorAction.REPORT)
	val byteBuffer = ByteBuffer.wrap(bytes)
	val result = ArrayList<Array<String?>>()
	val currentLine = ArrayList<String?>()
	var currentIndex = -1
	while (true) {
		val lastIndex = currentIndex
		while (true) {
			currentIndex++
			if (currentIndex >= bytes.size) {
				currentIndex = -1
				break
			}
			if (bytes[currentIndex].toUByte() >= 0xFE.toUByte()) {
				break
			}
		}
		val valueBytes = (byteBuffer.duplicate().position(lastIndex + 1).limit(if (currentIndex < 0) bytes.size else currentIndex) as ByteBuffer).slice()
		if (valueBytes.capacity() == 1 && valueBytes[0] == 0xFD.toByte()) {
			currentLine.add(null)
		} else if (valueBytes.capacity() == 1 && valueBytes[0] == 0xFC.toByte()) {
			currentLine.add("")
		} else if (valueBytes.capacity() >= 1) {
			try {
				currentLine.add(decoder.decode(valueBytes).toString())
			} catch (e: Exception) {
				throw RuntimeException("Invalid string value", e)
			}
		} else if (((currentIndex >= 0 && bytes[currentIndex] == 0xFF.toByte() || currentIndex < 0) && (lastIndex < 0 || lastIndex >= 0 && bytes[lastIndex] == 0xFF.toByte())) == false) {
			throw RuntimeException("Invalid BSV value byte sequence")
		}
		if (currentIndex < 0) {
			break
		} else if (currentIndex >= 0 && bytes[currentIndex] == 0xFF.toByte()) {
			result.add(currentLine.toTypedArray())
			currentLine.clear()
		}
	}
	result.add(currentLine.toTypedArray())
	return result.toTypedArray()
}
	
// ----------------------------------------------------------------------

@Throws(IOException::class)
fun saveBsv(jaggedArray: Array<Array<String?>>, filePath: String) {
	Objects.requireNonNull(filePath)
	Files.write(Paths.get(filePath), encodeBsv(jaggedArray))
}

@Throws(IOException::class)
fun loadBsv(filePath: String): Array<Array<String?>> {
	Objects.requireNonNull(filePath)
	return decodeBsv(Files.readAllBytes(Paths.get(filePath)))
}

@Throws(IOException::class)
fun appendBsv(jaggedArray: Array<Array<String?>>, filePath: String?) {
	var stream: OutputStream
	var existed: Boolean
	try {
		stream = Files.newOutputStream(Paths.get(filePath), StandardOpenOption.APPEND)
		existed = true
	} catch (exception: NoSuchFileException) {
		stream = Files.newOutputStream(Paths.get(filePath), StandardOpenOption.CREATE_NEW)
		existed = false
	}
	try {
		if (existed == true) {
			stream.write(0xFF)
		}
		stream.write(encodeBsv(jaggedArray))
	} finally {
		stream.close()
	}
}

// ----------------------------------------------------------------------

fun jaggedArrayToString(jaggedArray: Array<Array<String?>>): String {
	val sb = StringBuilder()
	sb.append("[")
	for (line in jaggedArray) {
		sb.append("\n  [")
		var isFirst = true
		for (value in line) {
			if (!isFirst) {
				sb.append(", ")
			}
			isFirst = false
			if (value == null) {
				sb.append("null")
			} else {
				sb.append("\"" + value.replace("\n", "\\n").replace("\u0000", "\\0") + "\"")
			}
		}
		sb.append("]")
	}
	sb.append("\n]")
	return sb.toString()
}
	
fun main(args: Array<String>) {
	try {
		val jaggedArray = arrayOf(arrayOf("Hello", "🌎", null, ""), arrayOf<String?>("A\u0000B\nC", "Test 𝄞"))
		println(jaggedArrayToString(jaggedArray))
		
		val bytes = encodeBsv(jaggedArray)
		val decodedJaggedArray = decodeBsv(bytes)
		println(jaggedArrayToString(decodedJaggedArray))
		
		saveBsv(jaggedArray, "Test.bsv")
		val loadedJaggedArray = loadBsv("Test.bsv")
		println(jaggedArrayToString(loadedJaggedArray))
		
		saveBsv(loadedJaggedArray, "TestResaved.bsv")
		
		appendBsv(arrayOf(arrayOf<String?>("ABC")), "Append.bsv")
	} catch (e: Exception) {
		println(e.toString())
	}
}